homeassistant/components/overkiz/coordinator.py

```diff
@@ -0,0 +1,189 @@
+    BadCredentialsException,
+    MaintenanceException,
+    NotAuthenticatedException,
+    TooManyRequestsException,
+)
+from pyoverkiz.models import DataType, Device, Place, State
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, UPDATE_INTERVAL
+
+DATA_TYPE_TO_PYTHON: dict[DataType, Callable[[DataType], Any]] = {
+    DataType.INTEGER: int,
+    DataType.DATE: int,
+    DataType.STRING: str,
+    DataType.FLOAT: float,
+    DataType.BOOLEAN: bool,
+    DataType.JSON_ARRAY: json.loads,
+    DataType.JSON_OBJECT: json.loads,
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OverkizDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching data from Overkiz platform."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        *,
+        name: str,
+        client: OverkizClient,
+        devices: list[Device],
+        places: Place,
+        update_interval: timedelta | None = None,
+        config_entry_id: str,
+    ) -> None:
+        """Initialize global data updater."""
+        super().__init__(
+            hass,
+            logger,
+            name=name,
+            update_interval=update_interval,
+        )
+
+        self.data = {}
+        self.client = client
+        self.devices: dict[str, Device] = {d.device_url: d for d in devices}
+        self.is_stateless = all(
+            device.device_url.startswith("rts://")
+            or device.device_url.startswith("internal://")
+            for device in devices
+        )
+        self.executions: dict[str, dict[str, str]] = {}
+        self.areas = self.places_to_area(places)
+        self._config_entry_id = config_entry_id
+
+    async def _async_update_data(self) -> dict[str, Device]:
+        """Fetch Overkiz data via event listener."""
+        try:
+            events = await self.client.fetch_events()
+        except BadCredentialsException as exception:
+            raise UpdateFailed("Invalid authentication") from exception
+        except TooManyRequestsException as exception:
+            raise UpdateFailed("Too many requests, try again later.") from exception
+        except MaintenanceException as exception:
+            raise UpdateFailed("Server is down for maintenance.") from exception
+        except TimeoutError as exception:
+            raise UpdateFailed("Failed to connect.") from exception
+        except (ServerDisconnectedError, NotAuthenticatedException):
+            self.executions = {}
+
+            # During the relogin, similar exceptions can be thrown.
+            try:
+                await self.client.login()
+                self.devices = await self._get_devices()
+            except BadCredentialsException as exception:
+                raise UpdateFailed("Invalid authentication") from exception
```

Comment:
In a future pr `raise ConfigEntryAuthFailed` to start a reauth flow

Reply:
👍🏻. We need to reimplement this since our current reauth flow is not well written, thus I left it out for the initial PR.
Good pointer.